File: 21551-naveen-siva-kumar/week4/329-longest-increasing-path-in-a-matrix/329-longest-increasing-path-in-a-matrix.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        
        if(matrix.empty()) return 0;
        
        int m = matrix.size(), n = matrix[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        int longest = 0;
        
        for(int i = 0; i < m; i++)
            for(int j = 0; j < n; j++){
                longest = max(longest, dfs(matrix, i, j, INT_MIN, dp));
            }
        
        return longest;
    }
    
    int dfs(const vector<vector<int>>& matrix, int i, int j, int prevEle, vector<vector<int>>& dp){
        
        if(i < 0 || i >= matrix.size() || j < 0 || j >= matrix[0].size()){
            return 0;
        }
        
        int top,bottom,left,right;
        
        if(matrix[i][j] > prevEle){
            
            if(dp[i][j]) return dp[i][j];
            
                top  = dfs(matrix, i - 1, j, matrix[i][j], dp) + 1;
            
                bottom  = dfs(matrix, i + 1, j, matrix[i][j], dp) + 1;
            
                left  = dfs(matrix, i, j - 1, matrix[i][j], dp) + 1;
            
                right = dfs(matrix, i, j + 1, matrix[i][j], dp) + 1;
            
            
            dp[i][j] = std::max(std::max(top, bottom), std::max(left, right));
            return dp[i][j];
        }
        
        return 0;
    }
    
};

File: 21560 - Sai Suraj/week 4/329_Longest Increasing Path in a Matrix.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        longest_by_pos={}
        directions=[(1,0),(-1,0),(0,1),(0,-1)]
        row,col = len(matrix),len(matrix[0])
        
        def dfs(pos):
            if pos in longest_by_pos:
                return longest_by_pos[pos]
            
            val=matrix[pos[0]][pos[1]]
            longest=1
            for dir_x,dir_y in directions:
                neighbor_x=dir_x+pos[0]
                neighbor_y=dir_y+pos[1]
                
                if 0<=neighbor_x<row and 0<=neighbor_y<col and matrix[neighbor_x][neighbor_y]>val:
                    longest=max(longest,1+dfs((neighbor_x,neighbor_y)))
            
            longest_by_pos[pos]=longest
            return longest
        
        max_path=1
        for r in range(row):
            for c in range(col):
                max_path=max(max_path,dfs((r,c)))
        return max_path
